scripts/wsrep_sst_clone.sh

```diff
@@ -0,0 +1,1113 @@
+
+    # parent process already did the master key rotation
+    INVARIANT_OPTIONS="--binlog-rotate-encryption-master-key-at-startup=OFF"
+
+    wsrep_log_debug "-> CLONE_SOCK $CLONE_SOCK "
+    wsrep_log_debug "-> CLONE_LIBS $CLONE_LIBS "
+
+
+    DEFAULT_OPTIONS=" \
+     $DEFAULTS_FILE_OPTION \
+     $DEFAULTS_GROUP_OPTION \
+     $INVARIANT_OPTIONS \
+     --datadir='$WSREP_SST_OPT_DATA' \
+     --wsrep_on=0 \
+     --secure_file_priv='' \
+     --socket="$CLONE_SOCK" \
+     --skip-mysqlx \
+     --log_error='$CLONE_ERR' \
+     --pid_file='$CLONE_PID_FILE' \
+     --plugin-dir='$CLONE_LIBS' \
+     --port='$JOINER_CLONE_PORT' \
+    "
+    SKIP_NETWORKING_OPTIONS=" \
+     --skip-networking \
+     --skip-mysqlx \
+    "
+
+    #define USER and PW
+    CLONE_PSWD=`wsrep_gen_secret`"$PSWD_POLICY"
+    CLONE_USER="clone_sst"
+
+##################################################################################################
+    #LET US Move the send of the READY here.
+    # 1) we send the message ready
+    # 2) we wait for donor to send message
+    # 3) IF message is SST@DONORIP:PORT (inside the IST file) we exit loop and continue SST
+    #    IF IST we exit SST script sending the UUID:position
+    # 4) IF SST Process continue
+
+    #check if there is another netcat process on the IP port we need and try to kill it.
+    NETCAT_KILL=`ps -xo pid,command|grep -e "nc -l -k $JOINER_CLONE_HOST $TSST_PORT"|grep -v "grep"|cut -d ' ' -f 1`
+    if [ ! "$NETCAT_KILL" == "" ];then
+        wsrep_log_info "-> Existing NetCat PID: $NETCAT_KILL. Will try to kill it"
+        kill -9 $NETCAT_KILL
+        if [ "$?" != "0" ]; then
+            wsrep_log_debug "-> Cannot kill the existing NetCat process PID $NETCAT_KILL"
+        fi
+    fi
+
+    #OPEN NETCAT to receive messages from DONOR (like IST)
+    wsrep_log_info "-> Opening NETCAT: nc -l $JOINER_CLONE_HOST $TSST_PORT"
+    nc -l $JOINER_CLONE_HOST $TSST_PORT > $WSREP_SST_OPT_DATA/XST_FILE.txt &
+    NC_PID=$!
+    wsrep_log_info "-> NETCAT PID $NC_PID"
+    if [ "$NC_PID" == "" ];then
+        wsrep_log_error "-> Cannot open Netcat at given port $JOINER_CLONE_HOST $TSST_PORT check if the port is already taken"
+        exit 1
+    fi
+
+    # Report clone credentials/address to the caller
+    wsrep_log_debug "-> ready passing string |$CLONE_USER:$CLONE_PSWD]$JOINER_CLONE_HOST:$JOINER_CLONE_PORT|"
+    echo "ready $CLONE_USER:$CLONE_PSWD]$JOINER_CLONE_HOST:$JOINER_CLONE_PORT"
+
+    #WAIT for Donor message
+    wsrep_log_debug "-> wait $JOINER_TIMEOUT_WAIT_XST"
+    while [ ! -s  $WSREP_SST_OPT_DATA/XST_FILE.txt ];do
+         if [ "$JOINER_TIMEOUT_WAIT_XST" == "0" ]; then
+            wsrep_log_error "TIMEOUT Waiting for the DONOR MESSAGE"
+            break ;
+         fi
+         sleep 1
+
+        ((JOINER_TIMEOUT_WAIT_XST-=1))
+    done
+
+    wsrep_log_debug "-> WAIT DIR DONOR MESSAGE DONE"
+
+    if [[ ! `cat $WSREP_SST_OPT_DATA/XST_FILE.txt` =~ "SST@" ]];then
```

Comment:
Here we have 2 messages.
SST or IST, there is NO other. So if not SST is IST, or as you fear a broken one. No need to add code to check the obvious (IST)

Reply:
@kamil-holubicki,  the check is already there because the reg expression checks for the message `SST` terminated by `@` that's it. We do not need another terminator like `<EOF>`. The only relevant part of the message is `SST` after that the GTID:POS received is a plus to resolve the position in case the recovery position is not able to recover it.

Reply:
There is synchronization issue here.
We are waiting in `while [ ! -s  $WSREP_SST_OPT_DATA/XST_FILE.txt ];do` loop. Once the file is not empty, we exit the loop. Note that the content of the file may be `SS` (not all data written to the file)
Then we go to `if [[ ! `cat $WSREP_SST_OPT_DATA/XST_FILE.txt` =~ "SST@" ]];then`. This `if` will detect that no full SST is needed, as the file contains only `SS`

About IST only. I don't see any place where donor sends the info about IST. It sends info only if full SST is needed.
The point is, that we have to wait in the loop until we receive the whole message, then we can parse this message.

Reply:
`where only one "s" is sent` - It is not about only "s" being sent. The full message is always sent. I agree. The point is about the receiver side. If you show me the justification that netcat writes data to the file in the atomic way, I will agree. For now, on the receiver side, you may have "SS" written to the file, while "T@" still being in the netcat buffer. Then sst script tests if the file is empty -> no it is not empty, because it contains "SS". Then the decision is made that we should go for IST, not SST, but it is wrong.
I agree that it might work in all tests, because there is a very short time window when we have only "SS" stored in file. This can lead to very hard to investigate bugs in the future. We can avoid this in one of the following ways:
1. Prove that netcat writes the whole message atomically into the file (I don't think it is the case)
2. Design the protocol in the way that protects us

As for bypass, you are right. I missed this line:
`echo $WSREP_SST_OPT_GTID | nc -w 1 $SST_HOST_STRIPPED $WSREP_SST_OPT_REMOTE_HOSTPORT || :`
so the joiner is informed.

Reply:
This PR fixes the above issue
https://github.com/Tusamarco/percona-xtradb-cluster/pull/2